### PR TITLE
[MRG] HW10

### PR DESCRIPTION
**Pull request by kjacks21**
# Description
**Goal**: Object class contracts.

As it happens, Liskov's implementation of clone() for the IntSet class (see figure 5.10, page 97) is wrong.

- Use the version of IntSet from the in-class exercise. Implement a subtype of IntSet to demonstrate the problem. Your solution should include appropiate executable code in the form of JUnit tests.
- Provide a correct implementation of clone() for IntSet. Again, give appropriate JUnit tests.
- Correctly override hashCode() and equals(). As discussed in the class exercise, the standard recipe is not appropriate in this (unusual) case.
- In addititon to code and tests, your deliverable is a story. Explain what is going on at each stage of the exercise. The GTA will primarily grade your story.

# Tasks

- [ ] Implement subtype of IntSet and demonstrate the problem

- [ ] Implement clone() and JUnit tests

- [ ] hashCode()

- [ ] equals()

- [ ] story

### CODE REVIEW

**Comment by 1000DS000000** [comment]
So should we split this hw up so that everyone does a bullet point and write their part of the story?
I can work on the part with clone method.

**Comment by kjacks21** [comment]
Sounds good to me. I won’t be able to touch this until tomorrow evening

**Comment by kjacks21** [comment]
Since we have to demonstrate the problem with the previous clone(), should we have like a `badClone()` method or do we need to have another `badIntSet.java` file/class?

**Comment by ngreenwald89** [comment]
I'll take `equals()`.

**Comment by ngreenwald89** [comment]
Not sure if I've done `equals()` right, and avoided the standard recipe. Please check my work.

**Comment by 1000DS000000** [comment]
I think the els==s.els should b els.equals(s.els) also. i think what they mean is to use getClass() instead of instanceof, because if we override it in a subclass, it wouldn't break the symmetry.

**Comment by kjacks21** [comment]
I'll do hashCode(). I've been busy all weekend and have today off from work, so I'll get to it shortly.

**Comment by 1000DS000000** [comment]
Are you also working on bullet point #1? If not, i can.

**Comment by kjacks21** [comment]
I am not currently working on it, go ahead (and thanks!). Just pushed up `hashCode` and tests.

File: hw10/src/IntSet.java [hunk 1]
```diff
@@ -12,8 +12,12 @@
 
 
     @Override public int hashCode() {
-        // see below
-        return 42;
+        int result = 0;
+        for (Integer i : els) {
+//            result += i.hashCode();  // from class
+            result += 31 * result + i.hashCode();  // bloch
```
Comment (kjacks21) [inline hw10/src/IntSet.java]: I figured bloch's recipe was better than the in class one. Thoughts?

**Comment by ngreenwald89** [comment]
@1000DS000000 updated equals.

**Comment by 1000DS000000** [comment]
so all that's left seems to be the "story" part for bullet point #3

**Comment by kjacks21** [comment]
@1000DS000000 to properly show how the old `clone()` fails, what does the grader need to do? Kist uncomment the original `clone` and comment out ours?

**Comment by kjacks21** [comment]
Great. Is this good for merge and submission then?

**Comment by 1000DS000000** [comment]
i think so

**Comment by kjacks21** [comment]
assignment has been submitted

**Comment by kjacks21** [comment]
Feedback from grader:

> Ok! You could have also mentioned that the original equals() method is wrong, because it violates reflexivity property (i.e. x.equals(x) should return true)

**Comment by ngreenwald89** [comment]
What's the grade for assignment 10? It didn't get uploaded into my grades.

**Comment by kjacks21** [comment]
10.5/10

**Pull request merged**